internal/cephfs/groupcontrollerserver.go

```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+		response.GroupSnapshot.Snapshots = append(response.GroupSnapshot.Snapshots, r.Snapshot)
+	}
+
+	return response, nil
+}
+
+// createSnapshotAddToVolumeGroupJournal creates the snapshot and adds the
+// snapshotID and volumeID to the volume group journal omap. If the freeze is
+// true then it will freeze the subvolumes and subvolume groups before creating
+// the snapshot and unfreeze them after creating the snapshot. If the freeze is
+// false it will call createSnapshot and get the snapshot details for the
+// volume and add the snapshotID and volumeID to the volume group journal omap.
+// If any error occurs other than ErrInProgress it will delete the snapshots
+// and undo the reservation and return the error.
+func (cs *ControllerServer) createSnapshotAddToVolumeGroupJournal(
+	ctx context.Context,
+	req *csi.CreateVolumeGroupSnapshotRequest,
+	freeze bool,
+	vgo *store.VolumeGroupOptions,
+	vgs *store.VolumeGroupSnapshotIdentifier,
+	cr *util.Credentials,
+	fsMap map[string]core.SubVolumeQuiesceClient,
+	secrets map[string]string) (
+	*[]csi.CreateSnapshotResponse,
+	error,
+) {
+	var err error
+	var resp *csi.CreateSnapshotResponse
+
+	defer func() {
+		if err != nil && !errors.Is(err, cerrors.ErrInProgress) {
+			err = cs.deleteSnapshotsAndUndoReservation(ctx, vgs, cr, fsMap, secrets)
+			if err != nil {
+				log.ErrorLog(ctx, "failed to delete snapshot and undo reservation: %v", err)
+			}
+		}
+	}()
+
+	responses := make([]csi.CreateSnapshotResponse, 0)
+	for _, volID := range req.GetSourceVolumeIds() {
+		if !freeze {
+			err = subvolumeQuiesceWithExpireTimeout(ctx, vgo.RequestName, fsMap)
+			if err != nil {
+				log.ErrorLog(ctx, "failed to quiesce subvolume with timeout: %v", err)
+
+				return nil, err
+			}
+		}
+		// Create the snapshot for the volumeID
+		clusterID := getClusterIDForVolumeID(fsMap, volID)
+		if clusterID == "" {
+			return nil, fmt.Errorf("failed to get clusterID for volumeID %s", volID)
+		}
+		req := &csi.CreateSnapshotRequest{
+			SourceVolumeId: volID,
+			Name:           vgs.FsVolumeGroupSnapshotName + "-" + volID,
+			Secrets:        secrets,
+			Parameters: map[string]string{
+				"clusterID": clusterID,
+			},
+		}
+
+		resp, err = cs.createSnapshotAndAddMapping(ctx, req, vgo, vgs)
+		if err != nil {
+			// Handle cleanup
+			log.ErrorLog(ctx, "failed to create snapshot: %v", err)
+
+			return nil, err
+		}
+		responses = append(responses, *resp)
+	}
+
+	if !freeze {
+		err = releaseSubvolumeQuiesce(ctx, vgo.RequestName, fsMap)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to freeze subvolume: %v", err)
+
+			return nil, err
+		}
+	}
+
+	return &responses, nil
+}
+
+// releaseSubvolumeQuiesce releases the quiesce of the subvolumes and subvolume
+// groups in the filesystems for the volumeID's present in the
+// CreateVolumeGroupSnapshotRequest.
+func releaseSubvolumeQuiesce(ctx context.Context,
+	requestName string,
+	fsMap map[string]core.SubVolumeQuiesceClient,
+) error {
+	releasing := false
```

Comment:
Naming `releasing` seems a little confusing, isn't it the opposite? Maybe use `inProgress` or something?